language/libra-vm/src/libra_vm.rs

```diff
@@ -499,24 +500,34 @@ pub fn txn_effects_to_writeset_and_events_cached<C: AccessPathCache>(
         ops.push((ap_cache.get_module_path(module_id), WriteOp::Value(blob)))
     }
 
-    let ws = WriteSetMut::new(ops)
-        .freeze()
-        .map_err(|_| VMStatus::Error(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))?;
-
     let events = effects
         .events
         .into_iter()
-        .map(|(guid, seq_num, ty_tag, ty_layout, val)| {
+        .map(|(guid, ty_tag, ty_layout, val)| {
             let msg = val
                 .simple_serialize(&ty_layout)
                 .ok_or_else(|| VMStatus::Error(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))?;
             let key = EventKey::try_from(guid.as_slice())
                 .map_err(|_| VMStatus::Error(StatusCode::EVENT_KEY_MISMATCH))?;
-            Ok(ContractEvent::new(key, seq_num, ty_tag, msg))
+            Ok(ContractEvent::new(key, ty_tag, msg))
         })
         .collect::<Result<Vec<_>, VMStatus>>()?;
 
-    Ok((ws, events))
+    for event in events {
```

Comment:
This is the key piece of code: putting an event inside the WriteSet. There is a big decision point here: where do we put the `TypeTag` that describes the structure of the event?